# PIDF is proportional, integrator, and filtered derivative, which sets how motors reach 
# a set position intelligently. Look at following link to see what each component does:
# https://commons.wikimedia.org/wiki/File:PID_Compensation_Animated.gif
# Besides how to tune the motors with gif above, I suggest some more research but to sum it up,
# you want to generally use P and then a tiny bit of I to tune motors and I Zone to limit range of I.
# You get to use Phoenix Motors' encoders that keeps track of position. If you use percent output,
# it just spins at a percentage of maximum voltage but with PID, you can have it reach the exact 
# spot you want consistently with efficient movement too
# To sum: PID = smart motors; PercentOutput = dumb motors


mBuffer: 10

wheelBuffer: 100

bScrewBuffer: 200000

trencherBuffer: 10000

# Current PID config for wheels
wheel_cfg:
  closedloopRamp: 1

  # position behind line
  initialPosition: 0

  # recorded position at deposit sieve
  desiredPosition: 500

  # difference between the initial and desired position
  positionDifference: 500

  # Motion magic
  motionCruiseVelocity: 5
  motionAcceleration: 10
  motionCurveStrength: 0

  percentOutput: .5

# Updated wheel PID for accurate encoders to record position needed for deposit autonomy
  slot0: {
    kP: 0.05,
    kI: 0.00000,
    kD: 0.0,
    kF: 0.0,

    # Limits how much I build up there is so the error doesn't become unstable
    maxIntegralAccumulator: 16000
  }

# Ballscrew configuration, the ballscrew is what extends the trencher to deposit and mine
# How I have it set up, the ballscrew has a zero position when it is all the way contracted
bscrew_cfg:

  # ballscrew is fully contracted
  drivePosition: 0

  # ballscrew is fully contracted
  depositPosition: 0

  # ballscrew is fully extended
  digPosition: -7000000

  # this is what is used for motion magic to control speeds
  motionCruiseVelocity: -40000
  motionCruiseVelocityDown: -10000
  motionAcceleration: 100000
  motionCurveStrength: 8

  percentOutput: -.15

  # PID
  slot0: {
    kP: .05,
    kI: 0.0000,
    kD: 0.0
  }

# Linear actuator configuration, this changes the angle the trencher is facing ex. extend to deposit
# or extend to dig. 0 is the default and fully contracted position
linact_cfg:

  # trencher is fully contracted
  drivePosition: 400

  # trencher is all the way extended
  depositPosition: 710 

  # trencher is mostly extended
  digPosition: 650 
  miliseconds: 1000

  # Motion magic
  motionCruiseVelocity: 10
  motionAcceleration: 10
  motionCurveStrength: 0

  percentOutput: .5

  # PID
  slot0: {
    kP: 20, 
    kI: 0.0002,
    kD: 0.0
  }

# Deposits gravel by moving bucket, full contraction is 0 and default, while extension is deposit position
bucket_cfg:
  # bucket is slightly extended
  drivePosition: 60

  # bucket is fully extended
  depositPosition: 850

  # bucket is fully contracted
  digPosition: 0

  # Motion magic
  motionCruiseVelocity: 8
  motionAcceleration: 10
  motionCurveStrength: 0

  percentOutput: .5

  # PID
  slot0: {
    kP: 20,
    kI: 0.0002,
    kD: 0.0
  }

# rotates scoops to collect gravel
trencher_cfg:
  # trencher is fully contracted
  drivePosition: 0 

  # trencher is all the way extended
  depositPosition: 0 

  # trencher is mostly extended
  digPosition: 0 

  # Motion magic
  motionCruiseVelocity: 5
  motionAcceleration: 10
  motionCurveStrength: 0
  
  percentOutput: .35

  cycle: 736000

  # PID
  slot0: {
    kP: .05, 
    kI: 0.000,
    kD: 0.0
  }
